packages/protocol/contracts/layer1/based2/AbstractInbox.sol

```diff
@@ -0,0 +1,350 @@
+import "./libs/LibCodec.sol";
+import "./libs/LibPropose.sol";
+import "./libs/LibProve.sol";
+import "./libs/LibVerify.sol";
+import "./IInbox.sol";
+import "./IPropose.sol";
+import "./IProve.sol";
+
+/// @title AbstractInbox
+/// @notice Acts as the inbox for the Taiko Alethia protocol, a simplified version of the
+/// original Taiko-Based Contestable Rollup (BCR) but with the tier-based proof system and
+/// contestation mechanisms removed.
+///
+/// Key assumptions of this protocol:
+/// - Block proposals and proofs are asynchronous. Proofs are not available at proposal time,
+///   unlike Taiko Gwyneth, which assumes synchronous composability.
+/// - Proofs are presumed error-free and thoroughly validated, with subproofs/multiproofs management
+/// delegated to IVerifier contracts.
+///
+/// @dev Registered in the address resolver as "taiko".
+/// @custom:security-contact security@taiko.xyz
+abstract contract AbstractInbox is EssentialContract, IInbox, IPropose, IProve, ITaiko {
+    // -------------------------------------------------------------------------
+    // Constructor
+    // -------------------------------------------------------------------------
+
+    constructor() EssentialContract() { }
+
+    // -------------------------------------------------------------------------
+    // External Functions
+    // -------------------------------------------------------------------------
+
+    /// @notice Initializes the contract with owner and genesis block hash
+    /// @param _owner The owner address
+    /// @param _genesisBlockHash The genesis block hash
+    /// @param _gasIssuancePerSecond The initial gas issuance per second
+    function init4(
+        address _owner,
+        bytes32 _genesisBlockHash,
+        uint32 _gasIssuancePerSecond
+    )
+        external
+        initializer
+    {
+        _init(_owner, _genesisBlockHash, _gasIssuancePerSecond);
+    }
+
+    /// @inheritdoc IPropose
+    function propose4(
+        bytes calldata _packedSummary,
+        bytes calldata _packedBatches,
+        bytes calldata _packedEvidence,
+        bytes calldata _packedTrans
+    )
+        external
+        override(I, IPropose)
+        nonReentrant
+        returns (I.Summary memory)
+    {
+        I.Summary memory summary = _validateSummary(_packedSummary);
+        I.Batch[] memory batches = LibCodec.unpackBatches(_packedBatches);
+        I.BatchProposeMetadataEvidence memory evidence =
+            LibCodec.unpackBatchProposeMetadataEvidence(_packedEvidence);
+        I.TransitionMeta[] memory transitionMetas = LibCodec.unpackTransitionMetas(_packedTrans);
+
+        I.Config memory config = _getConfig();
+        LibState.Access memory access = _getReadWrite();
+
+        // Propose batches
+        summary = LibPropose.propose(config, access, summary, batches, evidence);
+
+        // Verify batches
+        summary = LibVerify.verify(config, access, summary, transitionMetas);
```

Comment:
Before preconf, this might be a little bit tricky for proposers because of the `_config.cooldownWindow` and   they don't know when their transactions will be landed on L1.

Reply:
This protocol is designed with preconfirmation as a precondition, if the preconfer is uncertain when his tx will land on L1, revert protection must be enabled to avoid tx revert.